File: src/lib/game/costs.ts
/**
 * What upgrades cost, as the Master Optimizer works it out. Every "from -> to" is the sum of each level's cost
 * from `from` up to `to - 1`.
 *
 * - Gold enhance: ATK, HP and HP Recovery cost J·L⁴/1e8 + L³/1e4 + L(L−1)/2 a level, J by level band (×1.00001
 *   a level past 1,000,000, measured to the target); CRIT DMG (L⁴/1e4 + 10L² + 10L)/20; Death Strike damage
 *   L⁵/2000 + 98L³; Death Strike chance L⁴; CRIT chance from its table.
 * - Cubes: weapons and accessories floor(grade factor × Orr[L+1] / 10000), classes the same with the class
 *   grade's factor; spirits 30·(L+1)³ cubes and their crystal table.
 * - Companion passives: Stone and Emerald by level (level 99's cost for every level past 100).
 * - Relics: the expected cost of each attempt (5000/11) over its success rate.
 */

export type CostTables = {
  gold: { critChance: number[]; bands: { from: number; multiple: number }[] };
  cubes: { orr: number[]; weaponFactors: Record<string, number>; classFactors: number[] };
  spiritCrystals: number[];
  companionPassives: Record<string, Record<string, { stone: number[]; emerald: number[] }>>;
};

/** Σk, Σk², Σk³ and Σk⁴ for k = 1..n. */
const s1 = (n: number) => (n * (n + 1)) / 2;
const s2 = (n: number) => (n * (n + 1) * (2 * n + 1)) / 6;
const s3 = (n: number) => s1(n) ** 2;
const s4 = (n: number) => (n * (n + 1) * (2 * n + 1) * (3 * n * n + 3 * n - 1)) / 30;
const s5 = (n: number) => (n * n * (n + 1) ** 2 * (2 * n * n + 2 * n - 1)) / 12;
/** Σ over L = from..to-1 of a sum function. */
const range = (sum: (n: number) => number, from: number, to: number) => (to > from ? sum(to - 1) - sum(Math.max(0, from - 1)) : 0);

const MILLION = 1_000_000;
const PAST_MILLION_BAND = 5000;
const PAST_MILLION_RATE = 1.00001;

/** Gold for ATK, HP or HP Recovery from one level to another. */
export function statGold(tables: CostTables, from: number, to: number): number {
  if (to <= from) return 0;
  // J·L⁴/1e8 + L³/1e4 + L(L−1)/2, the J part per band.
  const flat = range(s3, from, to) / 1e4 + (range(s2, from, to) - range(s1, from, to)) / 2;
  let quartic = 0;
  const { bands } = tables.gold;
  for (let i = 0; i < bands.length; i += 1) {
    const start = Math.max(from, bands[i]!.from);
    const end = Math.min(to, bands[i + 1]?.from ?? MILLION);
    if (end > start) quartic += bands[i]!.multiple * range(s4, start, end);
  }
  // Past a million, each 5,000-level band multiplies J by 1.00001 a level, counted up to the target level.
  const last = bands[bands.length - 1]?.multiple ?? 1;
  for (let band = 1, start = MILLION; start < to; band += 1, start += PAST_MILLION_BAND) {
    const lo = Math.max(from, start);
    const hi = Math.min(to, start + PAST_MILLION_BAND);
    if (hi <= lo) continue;
    quartic += last * PAST_MILLION_RATE ** Math.min(to - MILLION, PAST_MILLION_BAND * band) * range(s4, lo, hi);
  }
  return quartic / 1e8 + flat;
}

/** Gold for CRIT DMG levels. */
export const critDamageGold = (from: number, to: number) => (range(s4, from, to) / 1e4 + 10 * range(s2, from, to) + 10 * range(s1, from, to)) / 20;
/** Gold for Death Strike damage levels. */
export const deathStrikeGold = (from: number, to: number) => range(s5, from, to) / 2000 + 98 * range(s3, from, to);
/** Gold for Death Strike chance levels. */
export const deathStrikeChanceGold = (from: number, to: number) => range(s4, from, to);

/** Sums a per-level table over from..to-1 (levels past its end cost what its last level costs). */
function tableSum(table: readonly number[], from: number, to: number) {
  let total = 0;
  for (let level = from; level < to; level += 1) total += table[Math.min(level, table.length - 1)] ?? 0;
  return total;
}

/** Gold for CRIT chance levels. */
export const critChanceGold = (tables: CostTables, from: number, to: number) => tableSum(tables.gold.critChance, from, to);

/** Cubes to level a weapon or accessory of this grade. */
export function gearCubes(tables: CostTables, grade: string, from: number, to: number): number {
  const factor = tables.cubes.weaponFactors[grade] ?? 0;
  let total = 0;
  for (let level = from; level < to; level += 1) total += Math.floor((factor * (tables.cubes.orr[level + 1] ?? 0)) / 10000);
  return total;
}

/** Cubes to level a class of this grade (1 for Trainee, 2 for Adventurer, ...). */
export function classCubes(tables: CostTables, grade: number, from: number, to: number): number {
  const factor = tables.cubes.classFactors[grade - 1] ?? 0;
  let total = 0;
  for (let level = from; level < to; level += 1) total += Math.floor((factor * (tables.cubes.orr[level + 1] ?? 0)) / 10000);
  return total;
}

/** Cubes and Mana Crystals to level a spirit. */
export function spiritCost(tables: CostTables, from: number, to: number): { cubes: number; crystals: number } {
  return { cubes: 30 * (to > from ? s3(to) - s3(from) : 0), crystals: tableSum(tables.spiritCrystals, from, to) };
}

/** Stones and Emeralds to level a companion's passive. */
export function passiveCost(tables: CostTables, companion: string, passive: string, from: number, to: number): { stones: number; emeralds: number } | null {
  const table = tables.companionPassives[companion]?.[passive];
  if (!table) return null;
  return { stones: tableSum(table.stone, from, to), emeralds: tableSum(table.emerald, from, to) };
}

/** A relic attempt's success chance at a level: 100% at 0, then 99.4% down 1.7% a level, 10% from level 54. */
export const relicSuccessRate = (level: number) => (level <= 0 ? 1 : level >= 54 ? 0.1 : (99.4 - 1.7 * (level - 1)) / 100);

/** The expected cost of relic levels: each level's attempt cost (5000/11) over its success chance. */
export function relicCost(from: number, to: number): number {
  let total = 0;
  for (let level = from; level < to; level += 1) total += 5000 / 11 / relicSuccessRate(level);
  return total;
}

File: scripts/optimizer/costs.py
"""Upgrade cost tables: what levelling gear, spirits, enhance stats and companion passives costs.

Equipment Data holds the cube tables ("Orr Cost", "Weapon Cost Factor", "Class Cost Factor") and the spirit
crystal table ("SPIRIT COST"); Gold Enhancement Data holds the gold multiplier bands for ATK, HP and HP Recovery
("Multiple" from each "Multiplier Starting Interval") and the "CRIT % COST TABLE"; Companions Data holds a
Stone and Emerald table per companion, a (Stone, Emerald) column pair per passive.
"""

from optimizer.workbook import MissingHeader, find_cell, number, text


def _level_table(sheet, title, value_header):
    """Values by level under `title`: a Level column with `value_header` next to it, level 0 first."""
    title_row, title_col = find_cell(sheet, title)
    header_row = title_row + 1
    if text(sheet.cell(header_row, title_col).value) != "Level":
        raise MissingHeader(f"{sheet.title}: no Level column under {title!r}")
    value_col = None
    for c in range(title_col + 1, title_col + 3):
        if text(sheet.cell(header_row, c).value) == value_header:
            value_col = c
    if value_col is None:
        raise MissingHeader(f"{sheet.title}: no {value_header!r} column under {title!r}")
    values = []
    row = header_row + 1
    while row <= sheet.max_row:
        level = number(sheet.cell(row, title_col).value)
        if not isinstance(level, (int, float)):
            break
        if int(level) != len(values):
            raise ValueError(f"{sheet.title} row {row}: expected level {len(values)} under {title!r}, found {level}")
        values.append(number(sheet.cell(row, value_col).value) or 0)
        row += 1
    if not values:
        raise MissingHeader(f"{sheet.title}: no levels under {title!r}")
    return values


def extract_cube_costs(equipment_sheet):
    """The cube tables: Orr base cost by level, the weapon/accessory factor by grade and the class factor by class grade."""
    orr = _level_table(equipment_sheet, "Orr Cost", "Cost")

    title_row, title_col = find_cell(equipment_sheet, "Weapon Cost Factor")
    weapon_factors = {}
    row = title_row + 2
    while row <= equipment_sheet.max_row:
        grade = text(equipment_sheet.cell(row, title_col).value)
        factor = number(equipment_sheet.cell(row, title_col + 1).value)
        if not grade or not isinstance(factor, (int, float)):
            break
        weapon_factors[grade] = factor
        row += 1

    title_row, title_col = find_cell(equipment_sheet, "Class Cost Factor")
    class_factors = []
    row = title_row + 2
    while row <= equipment_sheet.max_row:
        grade = number(equipment_sheet.cell(row, title_col).value)
        factor = number(equipment_sheet.cell(row, title_col + 1).value)
        if not isinstance(grade, (int, float)) or not isinstance(factor, (int, float)):
            break
        if int(grade) != len(class_factors) + 1:
            raise ValueError(f"{equipment_sheet.title} row {row}: expected class grade {len(class_factors) + 1}, found {grade}")
        class_factors.append(factor)
        row += 1

    if not weapon_factors or not class_factors:
        raise MissingHeader(f"{equipment_sheet.title}: empty cost factor tables")
    return {"orr": orr, "weaponFactors": weapon_factors, "classFactors": class_factors}


def extract_spirit_crystals(equipment_sheet):
    """Mana Crystals to take a spirit from each level to the next, level 0 first."""
    return _level_table(equipment_sheet, "SPIRIT COST", "Crystal")


def extract_gold_costs(gold_sheet):
    """CRIT % gold by level, and the ATK/HP/HP Recovery multiplier bands: [{from, multiple}] up to level 1,000,000."""
    crit_chance = _level_table(gold_sheet, "CRIT % COST TABLE", "COST TO NXT LEVEL")

    header_row, multiple_col = find_cell(gold_sheet, "Multiple")
    start_col = multiple_col + 1
    if text(gold_sheet.cell(header_row, start_col).value) != "Multiplier Starting Interval":
        raise MissingHeader(f"{gold_sheet.title}: no Multiplier Starting Interval next to Multiple")
    bands = []
    row = header_row + 1
    while row <= gold_sheet.max_row:
        multiple = number(gold_sheet.cell(row, multiple_col).value)
        start = number(gold_sheet.cell(row, start_col).value)
        if not isinstance(multiple, (int, float)):
            break
        start = 1 if not bands and not isinstance(start, (int, float)) else start
        if not isinstance(start, (int, float)) or start >= 1_000_000:
            break
        bands.append({"from": int(start), "multiple": multiple})
        row += 1
    if not bands:
        raise MissingHeader(f"{gold_sheet.title}: no gold multiplier bands")
    return {"critChance": crit_chance, "bands": bands}


def extract_companion_passive_costs(companions_sheet, companions):
    """{companion: {passive: {"stone": [by level 0-99], "emerald": [...]}}} from each companion's Level table."""
    costs = {}
    for companion in companions:
        # The companion's name sits over its table's Level column; the name appears elsewhere on the sheet too.
        found = None
        for row in companions_sheet.iter_rows():
            for cell in row:
                if text(cell.value) == companion and text(companions_sheet.cell(cell.row + 1, cell.column).value) == "Level":
                    found = (cell.row, cell.column)
                    break
            if found:
                break
        if not found:
            raise MissingHeader(f"{companions_sheet.title}: no Level table under {companion!r}")
        title_row, title_col = found
        header_row = title_row + 1
        passives = {}
        c = title_col + 1
        while c <= companions_sheet.max_column:
            name = text(companions_sheet.cell(header_row, c).value)
            if not name:
                break
            stone, emerald = [], []
            row = header_row + 2
            while row <= companions_sheet.max_row:
                level = number(companions_sheet.cell(row, title_col).value)
                if not isinstance(level, (int, float)):
                    break
                stone.append(number(companions_sheet.cell(row, c).value) or 0)
                emerald.append(number(companions_sheet.cell(row, c + 1).value) or 0)
                row += 1
            passives[name] = {"stone": stone, "emerald": emerald}
            c += 2
        if not passives:
            raise MissingHeader(f"{companions_sheet.title}: no passives under {companion!r}")
        costs[companion] = passives
    return costs
